proposals/0000-package-internals.rst

```diff
@@ -0,0 +1,148 @@
+   exposing internal modules, with something like ::
+
+    Could not load module 'Data.OldList';
+    it is a hidden module in the package 'base-4.15.1.0'.
+    If you wish to be able to access an internal module anyway, you may
+    specify '-expose-package-internals base-4.15.1.0'; however, be warned
+    that the API from internal modules may change between minor version releases.
+
+#. The `Haskell Package Versioning Policy <https://pvp.haskell.org/>`_ specifies that changes in
+   internal modules do *not* require a major version change.
+
+#. Haddock output does not change: only exposed modules are included in the documentation.
+   Users wishing to depend on internal modules are expected to consult the source code.
+
+#. For the next three releases of GHC, any import of a module exported from ``ghc-prim``
+   will emit a warning (controlled by ``-Wghc-prim-exports`` and part of ``-Wcompat``),
+   unless ``-expose-package-internals ghc-prim`` is specified. The warning will inform
+   users that all modules in ``ghc-prim`` will become internal after three releases.
+   When that happens, the warning itself will also be removed.
+
+
+Examples
+--------
+
+*Example.hs*\ ::
+
+  module Example (foldl) where
+  import Data.OldList   -- internal module in `base`
+
+::
+
+  > ghc -c Example.hs
+  Example.hs:2:1: error:
+    Could not load module 'Data.OldList';
+    it is a hidden module in the package 'base-4.15.1.0'.
+    If you wish to be able to access an internal module anyway, you may
+    specify '-expose-package-internals base-4.15.1.0'; however, be warned
+    that the API from internal modules may change between minor version releases.
+
+  > ghc -c Example.hs -expose-package-internals base
+  > ls
+  Example.hs Example.o Example.hi
+
+Effect and Interactions
+-----------------------
+
+1. All GHC users can now access internal modules from other packages, improving
+   the situations outlined in the Motivation_.
+
+#. The exports from ``ghc-prim`` are now protected; users can discover that
+   using names from ``ghc-prim`` is potentially troublesome.
+
+#. There may be a desire to integrate support for ``-expose-package-internals`` in ``.cabal``
+   files, but that is beyond the scope of this proposal.
+
+Costs and Drawbacks
+-------------------
+
+1. With this proposal, package authors now have no way to prevent access of
+   internal modules. However, the primary reason to have internal modules
+   is to present a stable, well-designed API to clients. Because internal
+   modules remain distinct from this public API, the public API is preserved.
+
+   Note that this does not pose a (new) security hazard, given that Template Haskell
+   can be used to access definitions from internal modules.
+
+Alternatives
+------------
+
+1. We could imagine a design where the command-line flag is per-module instead
+   of per-package, but that seems unnecessarily fine-grained.
```

Comment:
Again, `{-# OPTIONS_GHC ... #-}` would immediately give the ability to configure this per module (and not need cabal support).

Reply:
Interesting. I meant "per module" as meaning "per internal module", whereas you interpreted it as "per importing module". Either one works!